File: Tutorial4-8LayerImplementation2/src/main/java/com/test/tutorial4_8layerimplementation2/di/MainActivityModule.java
package com.test.tutorial4_8layerimplementation2.di;


import android.content.Context;

import com.test.tutorial4_8layerimplementation2.MainActivity;
import com.test.tutorial4_8layerimplementation2.di.model.ToastDisplayer;
import com.test.tutorial4_8layerlibrary.di.ActivityScope;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;

@Module
public abstract class MainActivityModule {

    /*
    NOTE: This can only be MainActivity(AndroidInjector<T>), Activity does not work
     */
    @Binds
    abstract Context bindContext(MainActivity activity);

    @ActivityScope
    @Provides
    public static ToastDisplayer provideActivityToastMaker(Context context) {
        return new ToastDisplayer(context);
    }
}

File: Tutorial5-4Dagger-Android-MultipleActivitiesAndFragments/src/main/java/com/smarttoolfactory/tutorial5_4dagger_android_multipleactivitiesandfragments/main/MainActivityModule.kt
package com.smarttoolfactory.tutorial5_4dagger_android_multipleactivitiesandfragments.main

import android.content.Context
import com.example.tutorial5_3dagger_android_multipleactivitiesandfragments.di.scope.ActivityScope
import com.example.tutorial5_3dagger_android_multipleactivitiesandfragments.model.ToastMaker
import dagger.Binds
import dagger.Module
import dagger.Provides

/**
 * Scope of @ContributesAndroidInjector methods and their modules should be same.
 * Otherwise app returns HAS CONFLICTING SCOPES error
 */
@Module
abstract class MainActivityModule {

    @Binds
    abstract fun bindContext(activity: MainActivity): Context

    @Module
    companion object {

        @JvmStatic
        @ActivityScope
        @Provides
        fun provideToastMaker(context: Context): ToastMaker {
            return ToastMaker(context)
        }
    }
}